tools/userland-mapping

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/python2.7
+
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2018 Adam Stevko
+#
+
+#
+# mapping.py - generate mapping between component FMRI and component package names, to be used for different purposes,
+# e.g. checking for outdated pacakges, vulnerable packages etc.
+#
+
+from __future__ import print_function, absolute_import
+
+import argparse
+import json
+import os
+import re
+import logging
+import subprocess
+import multiprocessing
+import sys
+
+try:
+    from scandir import walk
+except ImportError:
+    from os import walk
+
+logger = logging.getLogger('userland-mapping')
+
+COMONENT_MAPPING_FILENAME = 'mapping.json'
+
+
+def find_component_paths(path, subdir='components', debug=False):
+    expression = re.compile(r'.+\.p5m$', re.IGNORECASE)
+
+    paths = []
+    workspace_path = os.path.join(path, subdir)
+
+    for dirpath, dirnames, filenames in walk(workspace_path):
+        for name in filenames:
+            if expression.match(name):
+                paths.append(dirpath)
+                del dirnames[:]
+                break
+
+    return paths
+
+
+def generate_component_data(component_path):
+    result = []
+
+    proc = subprocess.Popen(['gmake', 'print-value-COMPONENT_NAME', 'print-package-names'],
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE,
+                            cwd=component_path)
+
+    for out in proc.stdout:
+        result.append(out.rstrip())
+
+    component_name = result[0]
+    component_fmris = result[1:]
+    component_relative_path = component_path.split(os.environ['WS_TOP'])[-1].replace('/', '', 1)
+
+    return component_fmris, component_name, component_relative_path
+
+
+def generate_userland_mapping(workspace_path, subdir='components'):
+    mapping = []
+
+    paths = find_component_paths(path=workspace_path, subdir=subdir)
+    pool = multiprocessing.Pool(processes=multiprocessing.cpu_count())
+    results = pool.map(generate_component_data, paths)
+
+    for component_fmris, component_name, component_relative_path in results:
+        for component_fmri in component_fmris:
+            mapping.append({'component_name': component_name,
+                            'component_fmri': component_fmri,
+                            'component_dir': component_relative_path})
+
+    component_mapping_file = os.path.join(workspace_path, subdir, COMONENT_MAPPING_FILENAME)
+    with open(component_mapping_file, 'w') as f:
+        f.write(json.dumps(mapping, sort_keys=True, indent=4))
+
+
+def main():
+    try:
```

Comment:
Will it fail also when -w switch is provided?

Reply:
Good catch. Maybe I can remove the check completely given that the main consumer is the build system and there is no point it running it outside of it.